skiko/src/awtMain/cpp/linux/swingRedrawer.cc

```diff
@@ -0,0 +1,185 @@
+    }
+
+    ~OffScreenBuffer() {
+        glXDestroyPbuffer(display, pbuffer);
+    }
+};
+
+class OffScreenTexture
+{
+public:
+    GLuint tex;
+    GLuint fbo;
+
+    ~OffScreenTexture() {
+        glDeleteFramebuffers(1, &fbo);
+        glDeleteTextures(1, &tex);
+    }
+};
+
+extern "C"
+{
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_makeOffScreenContext(
+        JNIEnv *env, jobject redrawer)
+    {
+        OffScreenContext* context = new OffScreenContext();
+        return toJavaPointer(context);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_disposeOffScreenContext(
+        JNIEnv *env, jobject redrawer, jlong contextPtr)
+    {
+        OffScreenContext* context = fromJavaPointer<OffScreenContext *>(contextPtr);
+        delete context;
+    }
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_makeOffScreenBuffer(
+        JNIEnv *env, jobject redrawer, jlong contextPtr, jlong oldBufferPtr, jint width, jint height)
+    {
+        OffScreenContext* context = fromJavaPointer<OffScreenContext *>(contextPtr);
+        OffScreenBuffer* oldBuffer = fromJavaPointer<OffScreenBuffer *>(oldBufferPtr);
+        
+        OffScreenBuffer* buffer;
+        if (oldBuffer == nullptr || oldBuffer->width != width || oldBuffer->height != height) {
+            if (oldBuffer != nullptr) {
+                delete oldBuffer;
+            }
+            buffer = new OffScreenBuffer(context, width, height);
+        } else {
+            buffer = oldBuffer;
+        }
+
+        return toJavaPointer(buffer);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_disposeOffScreenBuffer(
+        JNIEnv *env, jobject redrawer, jlong bufferPtr)
+    {
+        OffScreenBuffer* buffer = fromJavaPointer<OffScreenBuffer *>(bufferPtr);
+        delete buffer;
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_startRendering(
+        JNIEnv *env, jobject redrawer, jlong contextPtr, jlong bufferPtr)
+    {
+        OffScreenContext* context = fromJavaPointer<OffScreenContext *>(contextPtr);
+        OffScreenBuffer* buffer = fromJavaPointer<OffScreenBuffer *>(bufferPtr);
+
+        glXMakeCurrent(context->display, buffer->pbuffer, context->context);
+    }
+
+    JNIEXPORT void JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_finishRendering(
+        JNIEnv *env, jobject redrawer, jlong contextPtr)
+    {
+        OffScreenContext* context = fromJavaPointer<OffScreenContext *>(contextPtr);
+        glFinish();
+        glXMakeCurrent(context->display, None, nullptr);
+        glBindFramebuffer(GL_FRAMEBUFFER, 0);
+        glBindTexture(GL_TEXTURE_2D, 0);
+    }
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_swing_LinuxOpenGLSwingRedrawer_createTexture(
```

Comment:
Better name it as `bindTexture`/`unbindTexture` instead of create/dispose.

It not only creates a texture, but also changes the global buffer via `glBindFramebuffer`, to which OpenGl will draw.

Also, we should move `glBindFramebuffer(GL_FRAMEBUFFER, 0)` to `unbindTexture` from `finishRendering` to be symmetric